Tests/src/artofillusion/SceneTest.java

```diff
@@ -10,32 +10,768 @@
+      ObjectInfo so = scene.getObject("Cube");
+      
+      assertEquals(++sceneObjects, scene.getNumObjects());
+
+      assertEquals(2, so.getTracks().length);
+      
+      
+    }
+    
+  /**
+   * Test to add to scene new object other way
+   * Check that object created, added to scene and contains default tracks...
+   */
+  @Test
+    public void testAddObjectAsNewObjectInfo() {
+
+      int sceneObjects = scene.getNumObjects();
+      scene.addObject(new ObjectInfo(new Cube(1d,1d,1d), new CoordinateSystem(), "Cube"), (UndoRecord)null);
+      ObjectInfo so = scene.getObject("Cube");
+      
+      assertEquals(++sceneObjects, scene.getNumObjects());
+
+      assertEquals(2, so.getTracks().length);
+
+    }
+    
+  /**
+   * Test to add to scene new object other way
+   * Check that object created, added to scene and contains default tracks...
+   */
+  @Test
+    public void testAddObjectAsNewObjectInfoToGovenPos() {
+
+      int sceneObjects = scene.getNumObjects();
+      scene.addObject(new ObjectInfo(new Cube(1d,1d,1d), new CoordinateSystem(), "Cube"), scene.getNumObjects(), (UndoRecord)null);
+      ObjectInfo so = scene.getObject("Cube");
+      
+      assertEquals(++sceneObjects, scene.getNumObjects());
+
+      assertEquals(2, so.getTracks().length);
+
+    }
+    
+  /**
+   * Test to add to scene new object with Undo info
+   * Check that Undo record contains proper data to revert operation
+   */    
+  @Test
+    public void testAddObjectWithUndo()
+    {
+
+      UndoRecord ur = new UndoRecord(null, false);
+      
+      int sceneObjects = scene.getNumObjects();
+
+      scene.addObject(new Cube(1d,1d,1d), new CoordinateSystem(), "Cube", ur);
+      assertEquals(++sceneObjects, scene.getNumObjects());
+      assertNotNull(scene);
+      assertNotNull(ur.getCommands());
+      assertEquals(1, ur.getCommands().size());
+      assertTrue(ur.getCommands().get(0) == UndoRecord.DELETE_OBJECT);
+      
+    }
+    
+  /**
+   * Add material to scene.
+   * Check that material listener event is triggered
+   */
+  @Test
+    public void testAddMaterial(){
+
+      Material mat = new UniformMaterial();
+      listenerFireCount = 0;
+      
+      scene.addMaterialListener( new ListChangeListener() {
+        @Override
+        public void itemAdded(int index, Object obj)
+        {
+          listenerFireCount++;
+        }
+
+        @Override
+        public void itemRemoved(int index, Object obj)
+        {
+        }
+
+        @Override
+        public void itemChanged(int index, Object obj)
+        {
+        }
+      });
+      scene.addMaterial(mat);
+      assertEquals(1, listenerFireCount);
+    }
+    
+  /**
+   * Add material to scene at given position.
+   * Check that material listener event is triggered
+   * Check that material is inserted at expected position
+   */
+  @Test
+    public void testAddMaterialAtGivenPos()
```

Comment:
This method is failing with the current code. Please check your logic, and sanity-check by adding a material through the script console, and then retrieving and printing the material list.

Reply:
Both testAddMaterialAtGivenPos and testAddTextureAtGivenPos test cases fails because of code contract violation as discussed in this [thread](https://sourceforge.net/p/aoi/discussion/47784/thread/7aa9f80e/)

Reply:
Thats right. I'd forgotten.